QtImageViewer/VideoProcessor.cpp

```diff
@@ -1,20 +1,73 @@
 #include "VideoProcessor.hpp"
 
-#include <PolySyncVideo.hpp>
-#include <PolySyncDTCException.hpp>
-
 #include <QImage>
-#include <QPixmap>
-#include <iostream>
 
-using namespace std;
+
+VideoProcessor::VideoProcessor()
+    :
+    decoder( { PIXEL_FORMAT_MJPEG, 320, 240, 30 },
+             { PIXEL_FORMAT_RGB24, 320, 240, 30 } )
+{}
+
 
 VideoProcessor::~VideoProcessor()
 {
     disconnectPolySync();
 }
 
+
 void VideoProcessor::slotRun()
 {
     connectPolySync();
 }
+
+
+void VideoProcessor::initStateEvent()
+{
+    // Register as a listener for the message type that the publisher
+    // is going to send.  Message types are defined in later tutorials.
+    registerListener( getMessageTypeByName( "ps_image_data_msg" ) );
+}
+
+
+void VideoProcessor::messageEvent(
+        std::shared_ptr< polysync::Message > message )
+{
+    using namespace polysync::datamodel;
+    if( auto incomingMessage = getSubclass< ImageDataMessage >( message ) )
+    {
+        if( incomingMessage->getPixelFormat() == PIXEL_FORMAT_MJPEG )
+        {
+            auto buffer = decodeBuffer( incomingMessage->getDataBuffer() );
+            if( buffer.size() > 0 )
+            {
+                auto image = buildQImageFromFrameBuffer( buffer );
+                if( not image.isNull() )
+                {
+                    emit signalPixmap( QPixmap::fromImage( image ) );
+                }
+            }
+        }
+        else
+        {
+            std::cout << "Warning: unsupported pixel format type; "
+                         "this image data is not viewable\n";
+        }
+
+    }
+}
+
+
+std::vector< uchar > VideoProcessor::decodeBuffer(
+        const std::vector<uchar> & buffer )
+{
+    decoder.decode( buffer );
+
+    return decoder.getCopyOfDecodedBuffer();
```

Comment:
Is there any way to return a shared reference to the decoded data instead of copying it? Or do you think that is an unnecessary consideration for this example?